serverless-functions/src/functions/features/callback-and-voicemail/studio/wait-experience.protected.js

```diff
@@ -0,0 +1,340 @@
+ * We don't want callbacks or voicemails to contribute to abandoned task metrics.
+ *
+ * @param {*} context
+ * @param {*} task
+ * @param {*} cancelReason
+ */
+async function cancelTask(context, task, cancelReason) {
+  const newAttributes = {
+    ...task.attributes,
+    conversations: {
+      ...task.attributes.conversations,
+      abandoned: "Follow-Up",
+    },
+  };
+
+  return await TaskRouterOperations.updateTask({
+    context,
+    taskSid: task.sid,
+    updateParams: {
+      assignmentStatus: "canceled",
+      reason: cancelReason,
+      attributes: JSON.stringify(newAttributes),
+    },
+    attempts: 0,
+  });
+}
+
+/**
+ * Updates the call with callback or voicemail TwiML URL, and then cancels the ongoing call task with the appropriate reason and attributes.
+ *
+ * Much of the logic is the same for callback or voicemail, so we're using this single function to handle both.
+ *
+ * @param {*} context
+ * @param {*} isVoicemail
+ * @param {*} callSid
+ * @param {*} taskSid
+ * @returns
+ */
+async function handleCallbackOrVoicemailSelected(
+  context,
+  isVoicemail,
+  callSid,
+  taskSid
+) {
+  const domain = `https://${context.DOMAIN_NAME}`;
+  const twiml = new Twilio.twiml.VoiceResponse();
+
+  const cancelReason = isVoicemail
+    ? "Opted to leave a voicemail"
+    : "Opted to request a callback";
+  const mode = isVoicemail ? "record-voicemail" : "submit-callback";
+
+  const task = await fetchTask(context, taskSid);
+
+  // Redirect Call to callback or voicemail logic. We need to update the call with a new TwiML URL vs using twiml.redirect() - since
+  // we are still in the waitUrl TwiML execution - and it's not possible to use the <Record> verb in here. We piggyback on the same approach for callbacks,
+  // though technically these could be handled entirely in the waitUrl TwiML
+  const redirectUrl = `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=${mode}&CallSid=${callSid}&enqueuedTaskSid=${taskSid}`;
+  const result = await VoiceOperations.updateCall({
+    context,
+    callSid,
+    params: { method: "POST", url: redirectUrl },
+    attempts: 0,
+  });
+  const { success, status } = result;
+
+  if (success) {
+    //  Cancel (update) the task with handy attributes for reporting
+    await cancelTask(context, task, cancelReason);
+  } else {
+    console.error(
+      `Failed to update call ${callSid} with new TwiML. Status: ${status}`
+    );
+    twiml.say(
+      sayOptions,
+      "Sorry, we were unable to perform this operation. Please remain on the line."
+    );
+    twiml.redirect(
+      `${domain}/features/callback-and-voicemail/studio/wait-experience?mode=main-wait-loop&CallSid=${callSid}&enqueuedTaskSid=${taskSid}&skipGreeting=true`
+    );
+    return twiml;
+  }
+  return "";
+}
+
+exports.handler = async function (context, event, callback) {
+  const domain = `https://${context.DOMAIN_NAME}`;
```

Comment:
Is this needed here? Usually twiml infers the domain when doing redirects and such.

Reply:
It appears that it's only the <Play> that needs it really

Reply:
It appears that it's only the <Play> that needs it really

Reply:
Oh and a few other things seems to resolve relative URLs to https://webhooks.twilio.com. e.g. Gather's action:

https://webhooks.twilio.com/features/callback-and-voicemail/studio/wait-experience?mode=handle-initial-choice&CallSid=CAe3c5953de2dad8f7b39a0699b2440a55&enqueuedTaskSid=WT85415bd15f183d9fdac72205fea95c1b

Reply:
Ah, in that case, might as well leave this be.